---
title: "Untitled"
author: "TC"
date: "11/27/2020"
output: bookdown::pdf_document2
---



# RNAmrf, evolutional covariation aided RNA sequence alignment  


## Introduction

Sequences of proteins, nucleotides(DNAs and RNAs) are essential in the modern biology research. Interestingly, sequences of these molecules have undergone considerable changes during evolution, while preserving the folded structures and biological functions. It has now become a main task to detect homology from diversed sequences, which is usually achieved by alignment methods. Today, alignment methods have been used in fold recognition, function annotation, building phylogenetic trees, finding new motifs or non-coding RNAs, predicting structural and functional features.

There are two closely related alignment problems: pairwise alignment and multiple sequence alignment. Pairwise alignment aims to align two sequences, which can be efficiently solved by dynamic programming methods. Further development of pairwise alignment include the use of additional information from structure constraints and predicted features. 

The multiple sequence alignment problem aims to build an alignment for multiple homologous sequences. Several strategies have been developed: First, pairwise-based alignment methods can be generalized to progressively produce Multiple-sequence alignment, which starts by aligning two sequences and further adds other sequences into alignment. The alignment quality of these methods are determined by the order of the added sequences. Second, researchers have developed consistency scores to evaluate an MSA, and the goal is to iteratively optimize for alignment consistency. However, it is still difficult to build alignment greater than 1000 sequences using these method. Until recently, sequences has become abundant, making it highly desirable to construct larger MSA with thousands and even millions of sequences. For example, the tRNA family has about 1.4 million full sequences. Alignment methods based on the probablistic framework has been developed, as represented by HMMblits and infernal. A simple yet effective strategy to construct larger MSA is to align new sequences to a pre-compiled, manually curated seed alignment. This strategy is closely related to our work presented here.


In this work, we developed MRFalign, a new RNA sequence aligner based on the Markov Random Field(MRF) model. Compared with existing pseudoknot-alignment methods(locaRNA, MARNA), RNAmrf does not need to predict the secondary structures and perform pairwise alignment. Compared with covariance model, CM only allows for incorporating nested base-pairing interactions (pseudoknots are ignored), while MRF is more general, as it allows for the description of a fully connected covariation dependency.

We showed that RNAmrf outperforms cmalign in a synthetic alignment test and real RNA alignment tasks including pseudoknotted structure.


## Methods

### Markov Random Field model for RNA alignment

An RNA structure can be described by a probablistic graph model: nucleotides(A,U,G and C) are represented by nodes, interactions between nucleotides are represented by edges. The distribution of nucleotide types observed in a site is jointly determined by the site preference and coupling interactions.A markov random field (MRF) is introduced to disentangle direct and indirect coupling. It aims at inferring a statistical model $P(A_1,...,A_L)$ for the entire sequence space $A_1,...,A_L$. Given a RNA sequence ($A_1$,...$A_L$), the joint probability $P(A_1,...,A_L)$ in a Boltzmann distribution is described by

$$
P(A_1,...,A_L)=\frac{1}{Z}exp\left \{ {\sum_{i<j}{e_{ij}(A_i,A_j)}}+\sum_{i}{h_i(A_i)} \right \}
$$

where i and j are sequence index of nucleotides $A_i$ and $A_j$. The model parameters are fitted to the MSA. Various approaches are designed to infer the model, include message-passing, mean-field approximation, and pseudolikelihood model. Researchers have used the coupling term $e_{ij}$ to infer physical contacts with success. 

From a thermodynamics perspective, the probability of observing a sequence $A_1,...,A_L$ under a MRF model is determined by its energy. Similarly, we can see the sequence-to-MRF alignment problem as to find the alignment with the highest probability. Additional gap energy terms are required In order to calculate the free energy of a given alignment, but are not included in the MRF model. We used the affine gap penalty to penalize the creation of gaps in an alignment.

Inferring the global optimal alignment in a sequence-to-MRF alignment is an NP problem. Inspired by the map_align method, we sought out to develop a heuristic alignment approach: First, an initial alignment is guessed by using the one-body field accounting for field conservation, using dynamic programming; then the alignment is iteratively refined in search for alignment with lowest energy. 

### Alignment Accuracy Assessment and Test sets

The alignment accuracy is assessed by column alignment accuracy and base-pair alignment accuracy. The column accuracy is the percentage of correctly aligned bases (or base-pairs) in the MSA columns, comparing the test and reference MSA. This accuracy can be understood as sensitivity. Column accuracy is calculated for all reference columns and non-paired columns.
The base-pair alignment accuracy is calculated as the percentage of correctly aligned base-pairs comparing test and reference MSA. The base-pairs are from the pre-defined secondary structure in the reference alignment.

Additionally, the overall alignment quality is evaluated by the Sum-of-Pair Scores (SPS), and Structure Conservation Index (SCI). SPS reflects the average pairwise-alignment accuracy between two MSAs with a range between 0 and 1. SCI evaluates the consistency of an alignment. SCI will be high if sequences fold together equally well as if fold individually. SCI will be low if no concensus fold can be found.

The re-aligned MSAs are further evaluated in the contact prediction task by predicting contacts using DCA analysis.

#### Synthetic Alignment

A synthetic alignment dataset is constructed to compare the performance of MRFalign to cmalign, the state-of-the-art RNA aligner, in an extreme setting: starting with a predefined secondary structure that contains two pseudoknotted stems, all four types of nucleotides plus gap are distributed with a frequency of 1/4 on un-paired sites. For the eight base pairs, they are randomly selected from the three canonical base pair types (AU, CG, and GU). Random insertions of length 0-4 bases are inserted in the non-paired region. Similarly, each non-paired site has a probability of 0.1 to be deleted. Three secondary structure with pseudoknots are designed by hand. For each SS, three hundred random sequences are generated for a given SS to generate a reference MSA.

The secondary structure used to generate sequences are denoted below following a dot-bracket notation. Briefly, the second line denotes the reference index. Letter x stands for aligned positions and dot stands for the insertion of 0-4 bases. The third line denotes the secondary structure in dot-bracket annotation, where pairs are assigned by a pair of brackets or letter pairs in upper and lowercase(A and a).

```
>Synthetic-1
xxxxxxxxxxxxxxxxx....xxxxxxxxx
((((.AAAA..)))).........aaaa..

>Synthetic-2
xxxxx..xxxxx..xxxxxxxxxxxxxxxx
.AAAA...((((....aaaa....))))..

>Synthetic-3
xxxxxx....xxxxxxxxxxxxxxxxxxxxxx....xxxxxxxxxxxxxx
.AAAA..............((((..aaaa................)))).
```

For each reference MSA, a covariance model is built with cmbuild. Both synthetic reference MSA and the designed secondary structure are used as input. Then the covariance model is used to re-align the three hundred sequences. Similarly, a MRF model is trained on the synthetic MSA by GREMLIN. The MRF model is used to re-align the sequences with MRFalign.


#### Rfam Pseudoknot Families

A total of 37 seed alignment are retrieved from RFAM database. These seed alignments are selected from RNA families with pseudoknots. We only selected RNA families with length less than 200 due to long computation time for constructing the MRF model and alignment. The covariance models shipped with RFAM database was used to re-align the seed alignment. For MRF, the seed alignment is used to construct the MRF model by GREMLIN, then the MRF model is used to align the seed sequences.

The Code to build the covariance model and train the MRF model is provided below:

```
cmbuild --hand example.cm MSA.fasta

gremlin_cpp -alphabet rna -i MSA.afa -mrf_o example.mrf -o example.gremlinscore
```


## Results

### Synthetic Alignment Performance

Table \@ref(tab:tabsynth) shows the alignment performance of MRFalign and cmalign on three synthetic MSAs. Comparing the overall performance, MRFalign appears to have lower alignment quality as indicated by the consistent lower SPS and SCI scores compared with cmalign. However, MRFalign shows consistent accuracy in aligning base-pairs in the pseudoknot and non-pseudoknot stem. Cmalign has the highest alignment accuracy for the non-pseudoknotted stem (0.92~0.97) but fail to align the pseudoknots in Synthetic-1 and Synthetic-3 MSA with poor accuracy of 0.01 and 0.09, respectively. This is expected, as only the non-pseudoknot SS is incorporated in the CM model. 


The last three columns in Table \@ref(tab:tabsynth) denotes the sensitivity, precision(positive predictve value, PPV) and F-score for contact prediction using R-scape. Both sensitivity and precision are near perfect for MRFalign in all three tests, while cmalign generated MSA have lower precision and sensitivity. This result suggests that MRFalign can improve the downstream contact prediction performance. Figure \@ref(fig:synthcontact) provides visualization of the DCA covariation scores calculated using MSAs from reference, MRFalign and cmalign in the three synthetic test. In the reference and MRFalign (Figure \@ref(fig:synthcontact) A, B, D, E, G and H), both normal and pseudoknot stems have high covariance scores (GREMLIN score > 6), showing consistency between MRFalign MSA and the seed MSA. However, using cmalign generated MSA, the pseudoknotted base pairs shows weak and smear covariance signals (Figure \@ref(fig:synthcontact) C, F and I). In the synthetic-2 test, both MRFalign and cmalign has an accuracy of 0.53 in aligning the pseudoknot region. In synthetic-1 and synthetic-3, the pseudoknot pairs in the MRFalign MSA shows clear covariance signal, but the covariance signals are difficult to distinguish in the CMalign MSA(Figure \@ref(fig:synthcontact) C and I).

The alignment result on the synthetic RNA suggest that MRFalign outperforms cmalign in aligning RNA sequences containing pseudoknotted structures. The improvement is further confirmed in the contact prediction task. 


```{r tabsynth, echo=FALSE}
load("~/GIT/thesis_tc/data/synth_tab_sel.rda")

library(kableExtra)


kable(synth_tab_sel, booktabs = T,digits=2, label="tabsynth",
      caption= "Alignment accuracy of cmalign and RNAmrf on the synthetic MSA") %>%
  kable_styling() %>%
  add_header_above(c(" ", "Column Accuracy" = 2, "BP Accuracy" = 3,"Overall"=2,"Contact Pred"=3)) %>%
  pack_rows("Synthetic-1", 1, 2) %>%
  pack_rows("Synthetic-2", 3, 4) %>% 
  pack_rows("Synthetic-3", 5, 6)


# synth_tab=synth_tab[1:2,1:5]

# knitr::kable(synth_tab,caption="", booktab=TRUE, digits=2) 
# mytable=tab
# library(papaja)

```


```{r synthcontact, echo=FALSE, fig.cap="DCA covariance scores calculated from reference MSA, MRFalign-generated MSA and cmalign-generated MSA in the three synthetic MSA alignment test. DCA score is calculated by GREMLIN, Darker blue stands for higher covariation signal."}
# load(here::here("data/synth_contact.rda"))
load("~/GIT/thesis_tc/data/synth_contact.rda")
library(patchwork)
synth_contact 
```

### RFAM

Next, we tested MRFalign on Rfam families with pseudoknot structure.




## Discussion
